### PR TITLE
Added a test to demonstrate a bug where de-duping a tombstoned blob makes it unusable

### PR DESCRIPTION
**Pull request by agarwal-navin**
Added a test that demonstrates the following bug:
After an attachment blob with id say `/_blobs/uniqueId1` has been marked inactive / sweep ready / tombstone, if someone uploads a blob with the same content, the server will give back the same id `/_blobs/uniqueId1`. Now, adding this blob's handle or getting it will result in inactive / sweep ready / tombstone error today.
However, when a blob is uploaded, its lifetime should be the same whether it is de-duped or not. So, in the above scenario, there should be no inactive / sweep ready / tombstone error.

### CODE REVIEW

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 1]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
             assert.strictEqual(response?.status, 200, `Expecting a 200 response`);
             assert(container2.closed !== true, "Container should not have closed");
         });
+
+        /**
+         * Function that inverses the does not reject condition, i.e., asserts for reject instead. Used in these tests
+         * to demonstrate that because of a bug we are not getting the expected results. Once the bug is fixed, these
+         * asserts should start working as expected.
+         */
+        async function assertDoesNotRejectInverse(block: Promise<any>, message?: string | Error) {
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:277]: nit: Maybe `assertWronglyRejects`? Or just use `assert.rejects` directly and put a comment like `// BUG BUG BUG` haha

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 2]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
...
+            return assert.rejects(block, message);
+        };
+
+        /**
+         * This test validates that when blobs are de-duped in different containers, these containers can use these
+         * blobs irrespective of whether the original blob is tombstoned. Basically, after uploading a blob, a container
+         * should be able to use it the same way whether it was de-duped or not.
+         * Note: Currently this fails because the behavior is not as expected and so it's expecting an error.
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:285]: Are you going to remember to remove this when it's fixed?

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 3]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
...
+            // expires.
+            await delay(sweepTimeoutMs / 2);
+            const container2 = await loadContainer(summary1.summaryVersion);
+            const container2MainDataStore = await requestFluidObject<ITestDataObject>(container2, "default");
+            // Upload the blob and keep the handle around until the blob uploaded by first container is tombstoned.
+            const container2BlobHandle =
+                await container2MainDataStore._runtime.uploadBlob(stringToBuffer(blobContents, "utf-8"));
+
+            // Wait for sweep timeout so that the blob uploaded by the first container is tombstoned.
+            await delay(sweepTimeoutMs + 10);
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:317]: nit: `sweepTimeoutMs / 2 + 10` right?  What is SessionExpiry set to and will this trigger that? I guess not.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 4]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
...
+
+            // Send an op to update the current reference timestamp that GC uses to make sweep ready objects.
+            mainDataStore._root.set("key", "value");
+            await provider.ensureSynchronized();
+
+            // Summarize so that the tombstoned blob is now part of the summary.
+            const summary2 = await summarizeNow(summarizer);
+
+            // Load a container from this summary and upload a blob with the same content as the tombstoned blob.
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:326]: So this is the 3rd copy of the blob?  Why is this necessary?

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 5]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
...
+            // This blob will get de-duped but it should be fine to use it because from this container's perspective
+            // it uploaded a brand new blob.
+            const container3 = await loadContainer(summary2.summaryVersion);
+            const container3MainDataStore = await requestFluidObject<ITestDataObject>(container3, "default");
+
+            const container3BlobHandle =
+                await container3MainDataStore._runtime.uploadBlob(stringToBuffer(blobContents, "utf-8"));
+            // Ideally, this should not reject but currently it will because of a bug with how blob de-dup interacts
+            // with GC.
+            await assertDoesNotRejectInverse(container3BlobHandle.get(), "Container3 should be able to use the blob");
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:336]: Why can't we just use Container2?  I think I'd like to see an assert about what happens with Container2 when you try to get the blob handle.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 6]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
...
+
+            // Reference the blob in container2 and container3 which should be valid. There should not be any asserts or
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:338]:
```suggestion
            // Reference the blob in container2 and container3 which should be valid. There should not be any asserts
```

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts [hunk 7]
```diff
@@ -268,6 +268,80 @@ describeNoCompat("GC attachment blob tombstone tests", (getTestObjectProvider) =
...
+            // or errors logged in any container because of this.
+            container2MainDataStore._root.set("container2BlobHandle", container2BlobHandle);
+            container3MainDataStore._root.set("container3BlobHandle", container3BlobHandle);
+
+            await provider.ensureSynchronized();
```
Comment (markfields) [inline packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:343]: Why?

**Review by markfields — APPROVED** [review]
Do we have this bug tracked yet?

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 3 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:317]: yes, you are right. missed that part!

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:285]: Do you mean the comment? Maybe I should just not add it to begin with.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 1 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:277]: assertWronglyRejects is better.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:343]: The op from container2 will pass through because the blob is not tombstoned in it yet. However, processing it should fail in other containers (container3 and summarizer) where it is marked tombstone.
I will add a comment.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 4 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:326]: In container2, the blob is not tombstoned but in container3 it is tombstoned. It tests 2 different scenarios:
1. Container3 - The blob is tombstoned and can't even be retrieved.
2. Container2- The blob in not tombstoned. It can be retrieved but referencing it will result in an error in other containers.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 5 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:336]: The blob can be retrieved in container2 because it is not tombstoned there.

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:343]: Note that it does not fail currently because we don't have tombstone errors on revived which is added here - https://github.com/microsoft/FluidFramework/pull/13352

**Review by tyler-cai-microsoft — APPROVED** [review]

File: packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts (hunk 7 above)
Comment (agarwal-navin) [reply packages/test/test-end-to-end-tests/src/test/gc/gcTombstoneAttachmentBlobs.spec.ts:343]: @tyler-cai-microsoft FYI

**Comment by agarwal-navin** [comment]
> Do we have this bug tracked yet?

https://dev.azure.com/fluidframework/internal/_workitems/edit/2745
